perma_web/perma/utils.py

```diff
@@ -586,65 +601,91 @@ def get_wr_session(request):
         request.session['wr_username'] = new_username
         wr_username = new_username
         is_new_session = True
-    if new_cookie and wr_session_cookie != new_cookie:
-        request.session['wr_session_cookie'] = new_cookie
-        wr_session_cookie = new_cookie
+    if new_cookie and wr_private_session_cookie != new_cookie:
+        request.session['wr_private_session_cookie'] = new_cookie
+        wr_private_session_cookie = new_cookie
         is_new_session = True
 
+    return wr_username, wr_private_session_cookie, is_new_session
+
+def get_wr_session(request, is_private=False, collection_slug=None):
+    """
+    Initializes Webrecorder session or retrieves existing session.
+    """
+    if not is_private:
+        wr_username = DEFAULT_WR_USER
+        wr_session_cookie = get_wr_public_session(request)
+        is_new_session = False
+
+    else:
+        wr_username, wr_session_cookie, is_new_session = get_wr_private_session(request)
+
+    if not is_new_session and collection_slug:
+        is_new_session = not get_wr_collection_exists(wr_username, wr_session_cookie, collection_slug)
```

Comment:
I'm not sure about checking for the existence of a collection here, and overloading `is_new_session` to store the answer. `is_new_session` used to be a good shortcut for telling us that we definitely needed to create a new collection but with this new setup, it isn't anymore. Instead of optionally passing in a `collection_slug` to this method and then overloading `is_new_session` to inform us about the existence of the collection, let's do away with `is_new_session` entirely, and just handle collection existence in `models.Link.init_replay_for_user`. You already have a bit of checking/error-handing there for this:
```
        except WebrecorderException as wr:
            if wr.data == {'error': 'duplicate_name'}:
```
which `returns` for duplicate public collections and re-uploads duplicate private collections (we should probably choose one or the other and be consistent). That should be good enough, I'd think, without `is_new_session` OR an explicit api call to check for the existence a collection before attempting to create it. Do you agree?

As for whether to re-upload or not.... maybe we should make that decision based on some technique for detecting whether or not the warc was successfully uploaded the first time around. Either a) we should assume true, and collections should be deleted if Perma attempts an upload and fails; or b) we should be able to query and find out. My instincts say (a). How about you?

Reply:
Now there is only a single upload call. The `ensure_login` api returns if collection is empty or not and upload happens if it is empty. I suppose a delete/re-try option could still be added, though it will retry if the user refreshes the page as well (WR should not upload a partial collection)